lib/rubocop/cop/style/pattern_argument.rb

```diff
@@ -0,0 +1,74 @@
+# frozen_string_literal: true
+
+module RuboCop
+  module Cop
+    module Style
+      # This cop looks for all? / any? / none? / one? calls where the block
+      # has a single statement and === has been invoked on the receiver so the
+      # === argument can be used as the predicate method argument.
+      #
+      # @example
+      #   # bad
+      #   %w[foo bar foobar].any? { |e| /oo/.match?(e) }
+      #
+      #   [1, 2, 3].any? { |e| (1..10).include?(e) }
+      #
+      #   [1, 2, 3].any? { |e| Set[1, 2, 3] === e }
+      #
+      #   [1, 2, 3].none? do |e|
+      #     Integer === e
+      #   end
+      #
+      #   # good
+      #   %w[foo bar foobar].any?(/oo/)
+      #
+      #   [1, 2, 3].any?(1..10)
+      #
+      #   [1, 2, 3].any?(Set[1, 2, 3])
+      #
+      #   [1, 2, 3].none?(Integer)
+      #
+      class PatternArgument < Cop
+        include RangeHelp
+        extend TargetRubyVersion
+
+        minimum_target_ruby_version 2.5
+
+        def on_block(node)
+          pattern_argument_candidate?(node) do |method, pattern|
+            add_offense(node,
+                        location: method.loc.selector,
+                        message: format(MSG, method: method.method_name, pattern: pattern.source))
+          end
+        end
+
+        def autocorrect(node)
+          lambda do |corrector|
+            _, replacement = pattern_argument_candidate?(node).map(&:source)
+            replacement = "(#{replacement})" unless replacement.match?(/^\(.+\)$/)
```

Comment:
Tricky. I was thinking of checking for `begin_type`, but this looks pretty safe like this.

Reply:
I moved that to use `RuboCop::AST::NodePattern.match`, since ranges can be endless, beginless, floats, integers, etc. and anything enclosed in parentheses would match that query. I thought the best tool is to use NodePattern. And to avoid destructuring, first we'd have to check the replacement is `begin_type?`, then check the element inside is `range_type?`, isn't it?